application.cpp

```diff
@@ -36,22 +40,32 @@ class application_impl {
             ioctx.run();
          });
 
+         // after creating the thread for handling signals, we can block signals in the current thread
          sigset_t blocked_signals;
-         sigemptyset(&blocked_signals);
-         sigaddset(&blocked_signals, SIGINT);
-         sigaddset(&blocked_signals, SIGTERM);
-         sigaddset(&blocked_signals, SIGPIPE);
-         sigaddset(&blocked_signals, SIGHUP);
+         get_target_sigset(&blocked_signals);
          pthread_sigmask(SIG_BLOCK, &blocked_signals, nullptr);
-#endif
+      }
+
+      void get_target_sigset(sigset_t *blocked_signals) {
+         sigemptyset(blocked_signals);
+         sigaddset(blocked_signals, SIGINT);
+         sigaddset(blocked_signals, SIGTERM);
+         sigaddset(blocked_signals, SIGPIPE);
+         sigaddset(blocked_signals, SIGHUP);
       }
 
       ~application_impl() {
          if(_signal_catching_thread.joinable()) {
             _signal_catching_io_ctx->stop();
             _signal_catching_thread.join();
          }
+         
+         // need to unblock signals, otherwise next thread created will inherit blocked signals
+         sigset_t blocked_signals;
+         get_target_sigset(&blocked_signals);
+         pthread_sigmask(SIG_UNBLOCK, &blocked_signals, nullptr);
       }
+#endif
```

Comment:
`~application_impl()` was outside `#ifdef` before. Now it is only inside non-win32. Is this intended?

Reply:
Yes it is intentional. If there is nothing to do (as is the case on win32), then the implicit destructor is fine, no need to declare an empty destructor.